developer-docs-site/docs/move/move-on-aptos/cryptography.md

```diff
@@ -0,0 +1,186 @@
+
+As an example, a Move developer can implement the popular Boneh-Lynn-Shacham (BLS) signature scheme generically over **any** curve by using [type arguments](../../../move/book/functions#type-parameters) for the curve type in their implementation:
+
+```move
+use std::option;
+use aptos_std::crypto_algebra::{eq, pairing, one, deserialize, hash_to};
+
+/// Example of a BLS signature verification function that works over any pairing-friendly
+/// group triple `G1`, `G2`, `Gt` where signatures are in `G1` and PKs in `G2. Points are
+/// serialized using the format in `FormatG1` and `FormatG2` and the hashing method is 
+/// `HashMethod`.
+/// 
+/// WARNING: This example is type-unsafe and probably not a great fit for production code.
+public fun bls_verify_sig<G1, G2, Gt, FormatG1, FormatG2, HashMethod>(
+    dst:        vector<u8>,
+    signature:  vector<u8>,
+    message:    vector<u8>,
+    public_key: vector<u8>): bool
+{
+    let pk   = option::extract(&mut deserialize<G2, FormatG2>(&public_key));
+    let sig  = option::extract(&mut deserialize<G1, FormatG1>(&signature));
+    let hash = hash_to<G1, HashMethod>(&dst, &message);
+    
+    // Checks if $e(H(m), pk) = e(sig, g_2)$, where $g_2$ generates $\mathbb{G}_2$
+    return eq(
+        &pairing<G1, G2, Gt>(&msg_hash, &pk), 
+        &pairing<G1, G2, Gt>(&sig, &one<G2>())
+    );
+}
+```
+
+Using the `bls_verify_sig` _generic_ function from above, developers can verify BLS signatures over **any** of the supported (pairing-friendly) curves.
+For example, one can verify BLS signatures over BLS12-381 curves by calling the function above with the BLS12-381 marker types as its type arguments:
+
+```move
+use aptos_std::bls12381_algebra::{
+    G1, G2, Gt, FormatG1Compr, FormatG2Compr, HashG1XmdSha256SswuRo
+};
+
+// Aborts with code 1 if the BLS signature over the BLS12-381 curve fails to verify. 
+assert(
+    bls_verify_sig<G1, G2, Gt, FormatG1Compr, FormatG2Compr, HashG1XmdSha256SswuRo>(
+        dst, signature, message, public_key
+    ),
+    1
+);
+```
+
+For more use cases of the `crypto_algebra` module, check out some Move examples:
+
+1. [Verifying Groth16 zkSNARK proofs](#groth16-zksnark-verifier) over **any** curve 
+2. [Verifying randomness from the `drand` beacon](#verifying-randomness-from-the-drand-beacon)
+
+## Building powerful cryptographic applications in Move
+
+### Veiled coins
+
+The [`veiled_coin` example](https://github.com/aptos-labs/aptos-core/tree/main/aptos-move/move-examples/veiled_coin/sources) demonstrates how to use [the Ristretto255 modules from above](#ristretto255-arithmetic-in-move) to add a reasonable layer of privacy to coin balances and transactions. 
+
+Specifically, users can **veil** their balance, keeping it hidden from everyone, including validators.
+Furthermore, a user can send a **veiled transaction** that hides only the sent amount from everybody, including validators.
```

Comment:
Nit, but having "only sent" here makes me think the received amount isn't hidden. I think the next line makes it clear the system isn't anonymous.

Reply:
Sure